### ISSUE TITLE
Protobuf APIs & tooling

### ISSUE THREAD

**Issue by zepatrik** [ory/keto#311]
**Is your feature request related to a problem? Please describe.**

Currently we use `protoc` wich is quite limited. A better tool might be https://github.com/uber/prototool
It comes with linting.
Also interesting: https://github.com/googleapis/gapic-generator

We might want to use a gateway to avoid writing REST APIs.

**Comment by robinbraemer** [comment]
**Note:** uber/prototool is the "older" tool and bufbuild (also by prototool developers from uber) is the new exciting tool that should be used instead.
- https://github.com/bufbuild/buf

**Comment by robinbraemer** [comment]
Here is the first draft of the APIs: https://github.com/robinbraemer/keto/commit/a2176cc5783ebea78db921958f446cbc36173e60

**Comment by robinbraemer** [comment]
> Ops! Accidentally closed

**Comment by robinbraemer** [comment]
As RBAC and others are planned. I imagine having many different API services. So it might be better to organize `api/` like this:

```
keto/
   acl/
     node/v1/
     admin/v1/
     v1/
   rbac
     node/v1/
     admin/v1/
     v1/
```
...where the `node` package allows for node intercommunication APIs.

(ACL still of cause being the base layer for other models like RBAC.)

**Comment by robinbraemer** [comment]
## "Zookie" renaming suggestions for the Keto project

Since "Zookie" somewhat is a pretty Zanzibar-ish term, I'd like to give it a new name in Keto and that we can define our API appropriately.

**Short recap of what a Zookie is:**
> **2.4 API**
> Zookies in ACL read and check requests specify staleness bounds for snapshot reads, thus providing one of Zanzibar’s core consistency properties. We choose to use an **opaque cookie** instead of the actual timestamp to discourage our clients from choosing arbitrary timestamps and to allow future extensions.

- A Zookie is a "cookie", as in web terminology, replacing the first letter with a `Z` taken from Zanzibar

**This is what I came up with to give "Zookie" its new name in the Keto project:**
- `Kookie`
  - Pretty simple
  - Passes the "Is it a bad word in another language?" - test :D
- `Ketokie`
  - Reasonable
  - Language test: means "Cut away" in Hindi
    - Maybe a lucky coincidence, as zookies have something to do with snapshots and cut away newer information

@zepatrik @aeneasr

**Comment by zepatrik** [comment]
I kinda dislike `kookie` because people might confuse it with actual `cookie`'s and think we have a typo.
`ketokie` on the other hand is not recognizable as being derived from `cookie` and sounds more like "keto key".
I would actually be fine with `zookie` as the project is based on zanzibar.
More descriptive names I can think of: `object version`, `object state`, `requirements`, `bounds`

**Comment by robinbraemer** [comment]
Good points, staying with `zookie` is okay too.

Another API related topic is the.. 
## Read API

> 2.4.1 Read
Our clients read relation tuples to display ACLs or group
membership to users, or to prepare for a subsequent write.
A read request specifies one or multiple tuplesets and an optional zookie.
Each tupleset specifies keys of a set of relation tuples. The
set can include a single tuple key, or all tuples with a given
object ID or userset in a namespace, optionally constrained
by a relation name. With the tuplesets, clients can look up
a specific membership entry, read all entries in an ACL or
group, or look up all groups with a given user as a direct
member.
...

This deduces to 3(?) different query formats a client can make:
- look up a specific membership entry
- read all entries in an ACL or group
- look up all groups with a given user as a direct member

Have you put these in consideration in your first drafted `Query` message?
```protobuf
message ReadRelationTuplesRequest {
  message Query {
    // Optional
    RelationObject object = 1;
    // Optional
    string relation = 2;
    // Optional
    oneof subject {
      string user_id = 3;
      RelationUserSet user_set = 4;
    }
  }

  // All conditions are concatenated with
  // an OR operator. If any query matches
  // the relation will be in the response
  Query query = 1;
  int32 page = 2;
  int32 per_page = 3;
}
```

**Comment by tacurran** [comment]
@zepatrik @robinbraemer  re Zookie (aka hybrid form of cannabis) I feel like its important to differentiate what we do in our open source efforts from Google terminology. The reason is that if Keto becomes widely adapted it will be based on some similar concepts, as far as I understand the discussion so far, yet different in that it will be a cloud independent OSS system. I suggest that we use a new term instead of "Zookie" , a name based on session cookies, and not on consistency tokens. Ory's cloud native projects use the term "token" or "JWT" pronounced "jot" in many places. So why not create something else?
Here are some ideas:
1. Yotta (10^24) or (1000^8) - I think it would also be a scope for Yottas , a quadrillion units, and it would help us size everything accordingly. We think 2025 (0-24) will be the year Ory breaks out, and 4! is also a wonderful sequence.
2. Zoery (z03ry) has some Zo and Ory.  Zoe (life) and Ory. I like it more than Zookie...
3. Manna also has two fours (2 4s: for you, for me) and is symbolic for a nourishment |\/|a|\||\|a

**Comment by zepatrik** [comment]
Interesting suggestions @tacurran. This term will pop up in the documentation a lot, and therefore it should be recognizable. I would like it be as descriptive as possible so people get that it is some kind of token/cookie. The documentation will be easier to understand if you know that it works analogously to cookies. At least it was the case for me when reading the zanzibar paper.

**Comment by zepatrik** [comment]
Regarding the read API:
The current state was partly designed by @nickufer
From my understanding I would say that clients can read relation tuples by any combination of partial relation tuple content. Some example use cases I came up with:

1. **object**: display a list of rules for one object
2. **object & relation**: display all users/groups that have e.g. write access
1. **subject & relation**: display all groups a user belongs to/display all objects a user has access to
1. **object & relation & subject**: check whether the relation tuple exists already before writing it
1. **relation**: get all groups that have members; get all directories that have content

This is what the current state of the read API allows. More sophisticated queries require a proper query language and are out of scope for the moment. As a workaround, clients can use the read API multiple times and then compute the desired set locally.

**Comment by robinbraemer** [comment]
Re: "Zookie" renaming

I would also opt for choosing a **descriptive name** (@zepatrik) and create **something new** (@tacurran).

So to make the name descriptive, and since it really is an opaque consistency **token** for each content version, effectively a token to a **snapshot** of a relation tuple, just as described when Zookie gets first mentioned in the paper.

How about?:
- `snapshot_token` _( 2 words = bad )_
- **`snaptoken`** _( a mixture of both = short, new & descriptive? :D )_

**Comment by robinbraemer** [comment]
@zepatrik I'm now debating about the contents of a `ListRelationTuplesResponse`.

This will be the request message, so far, using the aforementioned `Query` message, the `snaptoken` (:D) and the [pagination fields](https://google.aip.dev/158) I proposed to you in the meeting.
```protobuf

message ListRelationTuplesRequest {
  // The query for listing relation tuples.
  // Clients can specify any optional field to
  // partially filter for specific relation tuples.
  //
  // Example use cases:
  //  - object only: display a list of all rules of one object
  //  - relation only: get all groups that have members; e.g. get all directories that have content
  //  - object & relation: display all subjects that have e.g. write relation
  //  - subject & relation: display all groups a subject belongs to/display all objects a subject has access to
  //  - object & relation & subject: check whether the relation tuple already exists, before writing it
  //
  message Query {
    // Optional.
    Object object = 1;
    // Optional.
    string relation = 2;
    // Optional.
    Subject subject = 3;
  }
  // All field constraints are concatenated
  // with a logical AND operator.
  // An unspecified field specifies a `*` wildcard match.
  Query query = 1;
  // Optional. The staleness bound of this
  // snapshot read.
  bytes snaptoken = 2;
  // Optional. The maximum number of
  // RelationTuples to return in the response.
  int32 page_size = 3;
  // Optional. A pagination token returned from
  // a previous call to `ListRelationTuples` that
  // indicates where this listing should continue from.
  string page_token = 4;
}
```

Now the current response message would be this, including the next page token.
But there is one significant scalability/latency problem with it.
```protobuf
// The response of a ReadService.ListRelationTuples rpc.
message ListRelationTuplesResponse {
  // The relation tuples matching the list request.
  repeated RelationTuple relation_tuples = 1;
  // Optional. A pagination token returned from a previous call to `ListRelationTuples`
  // that indicates where this listing should continue from.
  string next_page_token = 2;
}
```

The problem is, depending on the query, many, if not all, tuples in the `repeated` field may have the exact same `object` and/or `relation` and/or `subject` field.

Imagine a list request that queries all subjects of a given (object, relation). All `RelationTuple`s would have the same (object, relation) again and again, which hugely adds to the network bandwidth (serverside + clientside), page by page. Similar goes for other query combinations. In this example, only returning the subjects in the `RelationTuple`s in the response would be all that's needed, leaving the `object`, `relation` field blank.

Since the server can't know what the client explicitly is looking for, the response fields should be somehow specifiable upfront in the request message.

Since the client knows what he is requesting for and knows what he will extract from the response.
This deduces to be a "Partial response" as specified by this [AIP](https://google.aip.dev/157).

We have 2 design choices:
- Views - where we strictly enumerate possible response views
- Read masks - most flexible. I highly recommend to use it in this situation
  - I've used read masks pretty successfully before in other API designs as they are straightforward.

We would add this field to the request as to what the client wants to get expanded by the server (and we don't have to `SELECT` those fields on the serverside, saving bandwidth on serverside between the database.):
- `google.protobuf.FieldMask expand_mask`

Would that be okay to you? I think having this in, the current API definitions (#315) can be merged in already 🎉 and more specified a bit later.

**Comment by aeneasr** [comment]
> As RBAC and others are planned. I imagine having many different API services. So it might be better to organize `api/` like this:
> 
> ```
> keto/
>    acl/
>      node/v1/
>      admin/v1/
>      v1/
>    rbac
>      node/v1/
>      admin/v1/
>      v1/
> ```
> 
> ...where the `node` package allows for node intercommunication APIs.
> 
> (ACL still of cause being the base layer for other models like RBAC.)

Isn't RBAC just a userset? Would we really need another API for this?

**Comment by robinbraemer** [comment]
@aeneasr right, unless we want to have RBAC specific API methods like ListRoles, AddSubjectToGroup, etc. and abstract it away from ACL, which I would separate into the RBAC proto package. If we even want an RBAC API since it's just ACL at the end.

**Comment by zepatrik** [comment]
As it is not necessary for RBAC, we should really not include it right now. It could even be a separate service, but the complexity it abstracts away is minimal. Possibly best to do it in an SDK tbh. Lets first get everything working with the basic, un-opinionated ACL and add such stuff as decoration on top once the core is done :wink: :cake:

**Comment by zepatrik** [comment]
And for the same reasoning, lets move this whole zookie renaming discussion right before the fist release and rename it then. We can have discussions right now with people who read the paper but do not know our lingo.

**Comment by robinbraemer** [comment]
@zepatrik https://github.com/ory/keto/issues/288#issuecomment-728859429

**Comment by zepatrik** [comment]
Closing as the protobuf tooling is done. Will reference comments from here in new issues as appropriate.
